server/src/main/java/org/elasticsearch/cluster/routing/allocation/AllocationFailuresResetService.java

```diff
@@ -0,0 +1,56 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.cluster.routing.allocation;
+
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.cluster.ClusterChangedEvent;
+import org.elasticsearch.cluster.ClusterState;
+import org.elasticsearch.cluster.ClusterStateListener;
+import org.elasticsearch.cluster.ClusterStateUpdateTask;
+import org.elasticsearch.cluster.routing.allocation.command.AllocationCommands;
+import org.elasticsearch.cluster.service.ClusterService;
+import org.elasticsearch.core.SuppressForbidden;
+import org.elasticsearch.logging.LogManager;
+import org.elasticsearch.logging.Logger;
+
+/**
+ * A service that resets allocation failures counter and triggers reroute, when certain conditions change.
+ * For example cluster version changed or new nodes joined cluster.
+ */
+public record AllocationFailuresResetService(ClusterService clusterService, AllocationService allocationService)
+    implements
+        ClusterStateListener {
```

Comment:
I like the idea of a separate listener/service as it allows us to select changes (in case we have more scenarios like node join) that should trigger error count resets. 

The con of this approach is that it introduces one more cluster state update in a series of changes once node joins.
I wonder if we can do the same as part of the existing cluster state update? Possibly in `org.elasticsearch.cluster.routing.allocation.AllocationService#disassociateDeadNodes` as it is called when node joins or leave the cluster.

Reply:
I'm pretty sure this'll be fine, bearing in mind that shard allocation failures should be very rare these days. I do like the flexibility and clarity of having this logic in one place rather than sprinkling it across the node-join and node-left executors (and who knows where else in future).